File: backend/app/models/simulator.py
from pydantic import BaseModel, Field
from typing import Any


class SimulationRequest(BaseModel):
    scenario_type: str  # RATE_HIKE, FX_SHOCK, GEO_CONFLICT, TARIFF_WAR
    params: dict[str, Any] = {}


class PortfolioSimulationRequest(BaseModel):
    """이미지 기반 포트폴리오 시나리오 분석 요청.

    image_data가 있으면 OCR로 종목 추출 후 시뮬레이션.
    holdings가 있으면 OCR 없이 바로 시뮬레이션 (시나리오 변경 시 활용).
    """

    scenario_type: str
    params: dict[str, Any] = {}
    # 경로 A: 이미지 업로드 (첫 요청)
    image_data: str | None = Field(
        default=None, description="Base64 인코딩된 포트폴리오 이미지"
    )
    media_type: str = Field(
        default="image/png", description="이미지 MIME 타입"
    )
    # 경로 B: 이미 추출된 보유 종목 (시나리오 변경 시)
    holdings: list[dict[str, Any]] | None = Field(
        default=None, description="이미 추출된 보유 종목 리스트"
    )


class SimulationResult(BaseModel):
    scenario_type: str
    params: dict[str, Any]
    result: dict[str, Any]


class PortfolioSimulationResult(BaseModel):
    """포트폴리오 시뮬레이션 응답 — 추출된 holdings도 함께 반환."""

    scenario_type: str
    params: dict[str, Any]
    result: dict[str, Any]
    holdings: list[dict[str, Any]] | None = None

File: src/lib/api/simulator.ts
import { apiFetch } from "./client";

/* ── 일반 시나리오 ── */

interface SimulationRequest {
  scenario_type: string;
  params: Record<string, unknown>;
}

interface SimulationResult {
  scenario_type: string;
  params: Record<string, unknown>;
  result: Record<string, unknown>;
}

export async function runSimulation(
  data: SimulationRequest,
): Promise<SimulationResult> {
  return apiFetch<SimulationResult>("/simulator/analyze", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/* ── 포트폴리오 기반 시나리오 ── */

export interface PortfolioHolding {
  ticker: string;
  name: string;
  quantity?: number;
  avg_price?: number;
  current_price?: number;
  profit_loss_rate?: number;
  confidence?: number;
}

export interface PortfolioHoldingImpact {
  ticker: string;
  name: string;
  direction: "POSITIVE" | "NEGATIVE" | "NEUTRAL";
  magnitude: "LOW" | "MEDIUM" | "HIGH";
  reason: string;
  action: string;
}

export interface PortfolioSimResult {
  summary?: string;
  probability?: string;
  time_horizon?: string;
  impact?: Record<
    string,
    { direction: string; magnitude: string; description: string }
  >;
  portfolio_impact?: PortfolioHoldingImpact[];
  recommended_actions?: string[];
}

interface PortfolioSimulationRequest {
  scenario_type: string;
  params: Record<string, unknown>;
  image_data?: string;
  media_type?: string;
  holdings?: PortfolioHolding[];
}

interface PortfolioSimulationResponse {
  scenario_type: string;
  params: Record<string, unknown>;
  result: PortfolioSimResult;
  holdings: PortfolioHolding[] | null;
}

export async function runPortfolioSimulation(
  data: PortfolioSimulationRequest,
): Promise<PortfolioSimulationResponse> {
  return apiFetch<PortfolioSimulationResponse>("/simulator/analyze-portfolio", {
    method: "POST",
    body: JSON.stringify(data),
  });
}
